python-frontend/src/test/java/org/sonar/python/types/TypeInferenceTest.java

```diff
@@ -681,7 +681,19 @@ void user_defined_attributes() {
       "def f():",
       "  e = Foo()",
       "  e.attr.bit_length()"
-    ).type()).isEqualTo(anyType());
+    ).type()).isEqualTo(DECL_INT);
+  }
+
+  @Test
+  void user_defined_attributes_reassigned() {
+    assertThat(lastExpression(
+      "class Foo:",
+      "  attr: int",
+      "def f():",
+      "  e = Foo()",
+      "  e.attr = 'hello'",
+      "  e.attr"
+    ).type()).isEqualTo(DECL_INT);
```

Comment:
We'll probably want to consider to support field sensitive type inference eventually, to correctly infer that the type is `STR` here.

However, this may come with its own set of challenges and is out of scope for this ticket. I feel like this case should be discussed in our next sprints about type inference.